x-pack/legacy/plugins/apm/public/components/app/TransactionDetails/WaterfallWithSummmary/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts

```diff
@@ -69,75 +56,150 @@ interface IWaterfallItemBase {
+): IWaterfallTransaction {
   return {
-    id: transaction.transaction.id,
-    parentId: transaction.parent && transaction.parent.id,
-    serviceName: transaction.service.name,
-    name: transaction.transaction.name,
+    docType: 'transaction',
+    doc: {
+      id: transaction.transaction.id,
+      transaction,
+      errorsCount: errorsPerTransaction[transaction.transaction.id] || 0,
+      parentId: transaction.parent?.id,
+      serviceName: transaction.service.name,
+      timestamp: transaction.timestamp.us,
+      name: transaction.transaction.name
+    },
     duration: transaction.transaction.duration.us,
-    timestamp: transaction.timestamp.us,
     offset: 0,
-    skew: 0,
-    docType: 'transaction',
-    transaction,
-    errorCount: errorsPerTransaction[transaction.transaction.id] || 0
+    skew: 0
   };
 }
 
-function getSpanItem(span: Span): IWaterfallItemSpan {
+function getSpanItem(span: Span): IWaterfallSpan {
   return {
-    id: span.span.id,
-    parentId: span.parent && span.parent.id,
-    serviceName: span.service.name,
-    name: span.span.name,
+    docType: 'span',
+    doc: {
+      id: span.span.id,
+      span,
+      parentId: span.parent?.id,
+      serviceName: span.service.name,
+      timestamp: span.timestamp.us,
+      name: span.span.name
+    },
     duration: span.span.duration.us,
-    timestamp: span.timestamp.us,
+    offset: 0,
+    skew: 0
+  };
+}
+
+function getErrorItem(error: APMError): IWaterfallError {
+  return {
+    docType: 'error',
+    doc: {
+      id: error.error.id,
+      error,
+      message: error.error.log?.message || error.error.exception?.[0]?.message,
+      parentId: error.parent?.id,
+      serviceName: error.service.name,
+      timestamp: error.timestamp.us
+    },
     offset: 0,
     skew: 0,
-    docType: 'span',
-    span
+    duration: 0
   };
 }
 
+function getAgentMarks(transaction?: Transaction): IWaterfallAgentMark[] {
+  const agent = transaction?.transaction.marks?.agent;
+  if (!agent) {
+    return [];
+  }
+  return Object.entries(agent).map(([name, ms]) => ({
+    docType: 'agentMark',
+    doc: {
+      mark: name
+    },
+    offset: ms * 1000,
+    duration: 0,
+    skew: 0
+  }));
+}
+
 export function getClockSkew(
   item: IWaterfallItem,
   parentItem?: IWaterfallItem
 ) {
-  if (!parentItem) {
+  // parent should never be a type of agentMark
+  if (!parentItem || parentItem.docType === 'agentMark') {
     return 0;
   }
-
   switch (item.docType) {
     // don't calculate skew for spans. Just use parent's skew
     case 'span':
       return parentItem.skew;
-
+    // agentMark doesnt have parent-child relationship, no adjustment is needed
+    case 'agentMark':
+      return 0;
+    // when an error happened before its parent, the skew has to be calculated with the parent skew and duration.
+    // e.g.: An error with a parent that is the entry transaction and it doesnt have parent (skew = 0).
+    case 'error':
```

Comment:
Afaik the error will always have a `parent.id` property if it occurred inside a transaction/span and will point to either the transaction or span. Instead of calculating skew for the error separately, I think we should only calculate skew for the transaction, and the error should inherit this skew (like we do with spans):

```js
    case 'error':
      return parentItem.skew;
```

Reply:
I have found some cases that even returning the parent.skew, the offset of the error was still below zero. That's why I'm also calculating it for errors.

Reply:
This sounds like a bug. Errors that happen inside a transaction should be able to inherit the skew we calculate for that transaction.